src/hash.c

```diff
@@ -599,50 +599,55 @@ int InitUserphrase(struct ChewingData *pgdata, const char *path)
 
 int HashFileSeekToUserPhrase(struct ChewingData *pgdata, HASH_ITEM *pItem, FILE *fpHash)
 {
-    int len    = 0;
-    int result = 0;
-    char buf[FIELD_SIZE];
-    const char *pc;
-    HASH_ITEM *pTmpItem;
-
-    pTmpItem = ALC(HASH_ITEM, 1);
-    if (!pTmpItem)
-        return 0;
-
-    fseek(fpHash, strlen(BIN_HASH_SIG) + sizeof(pgdata->static_data.chewing_lifetime), SEEK_SET);
-
-    while (!feof(fpHash)) {
-        memset(buf, 0x00, FIELD_SIZE);
-        if (fread(buf, FIELD_SIZE, 1, fpHash) != 1)
-        {
-            free(pTmpItem);
-            return 0;
-        }
+    int fsize           = 0;
+    int iret            = 0;
+    int item_index      = 0;
+    int hdrlen          = 0;
+    int result          = 0;
+    char *seekdump      = NULL;
+    char *buf           = NULL;
+    HASH_ITEM *pItemTmp = NULL;
+
+    pItemTmp = ALC(HASH_ITEM, 1);
+    if (!pItemTmp)
+        goto error;
 
-        len = (int)(buf[16]);
-        pTmpItem->data.phoneSeq = ALC(uint16_t, len + 1);
+    fseek(fpHash, 0, SEEK_END);
+    fsize = ftell(fpHash);
 
-        pc = &(buf[17]);
-        while (len--)
-            pc += 2;
+    buf = ALC(char, fsize);
+    if (!buf)
```

Comment:
If `buf` equals to zero, you can directly return 0 without going to `error` and attempting to `free`.